### DISCUSSION TITLE
remoteproc remoteproc0: can't stop rproc

### DISCUSSION THREAD

**Discussion by yeliangchen2** [polarfire-soc/polarfire-soc-documentation#304] [HSS and U-Boot]
linux version is following:
![image](https://user-images.githubusercontent.com/16462894/229387347-59ef2b4a-8384-4af3-902d-504782b4a06a.png)
HSS version is following:
![image](https://user-images.githubusercontent.com/16462894/229387407-2fa4db37-30dd-4f72-bcee-1dcbede82cec.png)
print the following errors while starting U54_4(runing bare matal) :
![image](https://user-images.githubusercontent.com/16462894/229387563-e829edb9-84cf-40af-b280-d204838b61ba.png)

**Comment by yeliangchen2** [comment]
starting U54_4 is ok,but stoping U54_4 is failed

**Reply by vfalanis** [reply]
Hi @yeliangchen2 ,

Based on the info provided, it looks like you are using HSS 2023.02 with kernel version 2022.11. The kernel version, polarfire-soc-amp-examples and HSS versions must be in sync (all using 2023.02) as they are not backwards compatible with 2022.11 release.

We always recommend to keep both the reference design and Linux image versions in sync when updating to a new release.

Could you please try using the linux4microchip+fpga-2023.02 kernel tag?

Thanks

**Reply by yeliangchen2** [reply]
hello, i have try it ,as follows.but can not be solved :
![image](https://user-images.githubusercontent.com/16462894/229479233-5917f248-baf5-461e-96ff-c41c1b300a14.png)
![image](https://user-images.githubusercontent.com/16462894/229482079-e423b047-6532-4c20-a3ee-2f54f7a11c20.png)

**Comment by yeliangchen2** [comment]
hi, based on HSS projects, could you provide me some documents about start and stop harts?  in other words ,I should how to start U54_4 and stop U54_4 in the HSS projects. (ps:  I finish the comunication of  linux and HSS)

**Reply by vfalanis** [reply]
Hi,

Were you able to solve the issue posted above? I believe you are using a custom AMP elf application, is that correct?  

It's difficult to tell what the issue you are facing is given the information provided. Have you tried testing the default mpfs-rpmsg-remote.elf application from the 2023.02 release? You should be able to start and stop the firmware from Linux with that default elf file provided in the release.

> could you provide me some documents about start and stop harts?

With regards this question, could you please elaborate on what you would like to achieve? The AMP elf application should be started/stopped from Linux using the remoteproc framework.

**Reply by yeliangchen2** [reply]
Hi,as you mentioned above,I was using a custom elf application.could you tell me the communication process between linux,hss and elf during linux starting U54 hart and stoping U54 hart. So that I can modify the custom elf according to that you said.

**Reply by vfalanis** [reply]
Hi,

Based on the screenshots you provided, I can see you are using the late boot mode of remoteproc. Is this correct? If so I'll briefly summarise the communication process.

**Late Boot Mode**: This mode can be used to load and start the remote context firmware from Linux using the Remoteproc framework.
This mode can be enabled when using the skip-autoboot feature in the hss-payload-generator.

An example HSS payload configuration file for late boot mode is provided [here](https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/applications-and-demos/asymmetric-multiprocessing/remoteproc.md#configure-the-hss-payload-for-late-boot-mode).

When the HSS starts, it reads the payload and detects if the skip-autoboot is set for a particular hart/context. If skip-autoboot flag is found it will stop the HSS from loading the firmware at boot time and let Linux do it at a later time instead.

Once Linux boots, the sysfs interface can be used to load firmware and start/stop the hart associated to the remote context. 

To load firmware you can use the following command:
```
echo -n <firmware_name.elf> > /sys/class/remoteproc/remoteproc0/firmware
```
Once the firmware is loaded, you can start the firmware on the remote context with the following command
```
echo start >/sys/class/remoteproc/remoteproc0/state
```

Optionally you can stop the firmware with the following command:
```
echo stop >/sys/class/remoteproc/remoteproc0/state
```

When using all of the commands shown above, the remoteproc drivers will automatically send start/stop request to the HSS, so there is nothing to be done in this end. You just need to make sure you have CONFIG_SERVICE_OPENSBI_RPROC=y enabled in the HSS. This is enabled by default in the Icicle Kit defconfig.

From a Bare Metal point of view, the key is to have the following things in your custom bare metal application:

1. **IMPORTANT**: Make sure you are compiling the application with remoteproc enabled. For example if compiling in RPMsg remote mode:
 make REMOTE=1 **REMOTEPROC=1** CROSS_COMPILE=${TARGET_PREFIX} EXT_CFLAGS='${EXT_CFLAGS}' 

This is important because when using the SoftConsole Remote and Master build configurations to compile the application, this flag is not set by default. 

2. You also need to make sure that the MIV_RP_MBOX_STOP message handling is ported into your application. If you based your custom elf application on the polarfire-soc-examples repository, you should have it already:

- https://github.com/polarfire-soc/polarfire-soc-amp-examples/blob/main/mpfs-rpmsg-bm/src/middleware/remoteproc/remoteproc.c#L129

- https://github.com/polarfire-soc/polarfire-soc-amp-examples/blob/main/mpfs-rpmsg-bm/src/middleware/rpmsg/rpmsg_lite/porting/platform/microchip/miv/rpmsg_platform.c#L243

More information on remoteproc can be found in our [documentation](https://mi-v-ecosystem.github.io/redirects/asymmetric-multiprocessing_remoteproc).

Hope this helps.

**Reply by EricZhangjie** [reply]
Hi,
 The MIV_RP_MBOX_STOP message handling is rproc_stop function
![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/109322517/10f11acc-e377-4954-b254-d73ff9aa63d5)
what does line116 "((func_t)SCRATCHPAD_START)();" used for? Is it the real purpose of stop the hart? SCRATCHPAD_START is a macro that is defined as the start address of scratchpad MEMORY section showed in link file(mpfs-ddr-loaded-by-boot-loader-remote.ld)

**Reply by vfalanis** [reply]
Hi @EricZhangjie,

It is used to jump to address 0x0A000000, which is the scratchpad start address. Once in scratchpad, the hart associated with the remote context  will be put in wait for interrupt (wfi) mode until the hart is started again using remoteproc.

**Comment by vfalanis** [comment]
Hey @yeliangchen2, I'm closing this discussion since there hasn't been any activity in around 4 weeks. I hope the information provided above resolved your question. Feel free to open it back up if you need any further clarification.
Thanks
